src/KiotaClientMockExtensions.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System.Linq.Expressions;
+using System.Reflection;
+using Microsoft.Kiota.Abstractions;
+using Microsoft.Kiota.Abstractions.Serialization;
+using NSubstitute;
+
+namespace Gainsway.Kiota.Testing;
+
+/// <summary>
+/// Provides extension methods for mocking Kiota-generated client classes and their responses.
+/// </summary>
+public static class KiotaClientMockExtensions
+{
+    /// <summary>
+    /// Creates a predicate expression to match a <see cref="RequestInformation"/> object
+    /// based on its URL template ending with the specified value.
+    /// </summary>
+    /// <param name="urlTemplate">The URL template to match.</param>
+    /// <returns>An expression that evaluates to true if the URL template matches.</returns>
+    private static Expression<Predicate<RequestInformation>> RequestInformationUrlTemplatePredicate(
+        string urlTemplate
+    ) => req => req.UrlTemplate!.EndsWith(urlTemplate);
+
+    /// <summary>
+    /// Creates a mock of the client class.
+    /// This is a generic method that can be used to create a mock of any Kiota generated client class.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public static T GetClientMock<T>()
+        where T : class
+    {
+        IRequestAdapter _clientAdapter;
+        _clientAdapter = Substitute.For<IRequestAdapter>();
+
+        return (T)Activator.CreateInstance(typeof(T), _clientAdapter);
```

Comment:
I would suggest to rename _clientAdapter to _requestAdapterMock as it better reflects its purpose. 

Perhaps the name of the method is also bit misleading, as we are not really creating a mock of the client, we are creating genuine client with a mocked IRequestAdapter implementation passed into its constructor.  But I don't feel as strongly about this point, as the purpose is to mock the client's behaviour in general.

Since we are dealing with Kiota specifically, the clients we are working here with  seem to inherit from BaseRequestBuilder (which then gets the IRequestAdapter passed in). We could then be more correct by using `public static T GetClientMock<T>() where T: BaseRequestBuilder` rather then generic `class` which would be unlikely to have required constructor method.

Reply:
Instance can be returned as null, which is not desired here:

```suggestion
    public static T GetClientMock<T>()
        where T : class
    {
        IRequestAdapter _requestAdapterMock;
        _requestAdapterMock = Substitute.For<IRequestAdapter>();

        var instance = Activator.CreateInstance(typeof(T), _clientAdapter);
        return instance as T
            ?? throw new InvalidOperationException(
                $"Unable to create an instance of {typeof(T).Name}."
            );
```